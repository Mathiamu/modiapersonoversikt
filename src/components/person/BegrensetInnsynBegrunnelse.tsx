import * as React from 'react';
import { BegrensetInnsynTyper } from '../../models/person/person';

interface Props {
    begrunnelseType: BegrensetInnsynTyper;
}

function BegrensetInnsynBegrunnelse({begrunnelseType}: Props) {
    switch (begrunnelseType) {
        case BegrensetInnsynTyper.Kode6:
            return <>Bruker har kode 6 adresse</>;
        case BegrensetInnsynTyper.Kode7:
            return <>Bruker har kode 7 adresse</>;
        case BegrensetInnsynTyper.EgenAnsatt:
            return <>Bruker faller inn under begrensning for egen ansatt</>;
        default:
            return <>Ukjent sikkerhetsbegrensing</>;
    }
}

export default BegrensetInnsynBegrunnelse;