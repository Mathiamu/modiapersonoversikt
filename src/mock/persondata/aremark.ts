import {
    AdresseBeskyttelse,
    EgenAnsatt,
    ForelderBarnRelasjon,
    ForelderBarnRelasjonRolle,
    FullmaktsRolle,
    Kjonn,
    LocalDate,
    LocalDateTime,
    Person,
    PersonStatus,
    SikkerhetstiltakType,
    SivilstandType,
    Skifteform
} from '../../app/personside/visittkort-v2/PersondataDomain';

export const barnAremark: ForelderBarnRelasjon[] = [
    {
        ident: '12345678910',
        rolle: ForelderBarnRelasjonRolle.BARN,
        navn: [
            {
                fornavn: 'BARN1',
                mellomnavn: null,
                etternavn: 'BARNESEN'
            }
        ],
        fodselsdato: ['2008-03-15' as LocalDate],
        alder: 13,
        adressebeskyttelse: [
            {
                kode: AdresseBeskyttelse.UGRADERT,
                beskrivelse: 'UGRADERT'
            }
        ],
        bostedAdresse: [
            {
                linje1: 'Islandsgate 49',
                linje2: '7693 Hovedøya',
                linje3: 'Norge'
            }
        ],
        personstatus: [
            {
                kode: PersonStatus.BOSATT,
                beskrivelse: 'BOSATT'
            }
        ]
    }
];

export const aremark: Person = {
    fnr: '10108000398',
    navn: [
        {
            fornavn: 'AREMARK',
            mellomnavn: null,
            etternavn: 'TESTFAMILIEN'
        }
    ],
    kjonn: [
        {
            kode: Kjonn.M,
            beskrivelse: 'Mann'
        }
    ],
    fodselsdato: ['1980-10-10' as LocalDate],
    alder: 41,
    dodsdato: [],
    bostedAdresse: [
        {
            linje1: 'Islandsgate 49',
            linje2: '7693 Hovedøya',
            linje3: 'Norge'
        }
    ],
    kontaktAdresse: [
        {
            linje1: 'Aremarkveien 1',
            linje2: '0320 Aremark',
            linje3: null
        }
    ],
    navEnhet: {
        id: '0219',
        navn: 'NAV Bærum'
    },
    statsborgerskap: [
        {
            land: {
                kode: 'NOR',
                beskrivelse: 'NORGE'
            },
            gyldigFraOgMed: '1980-10-10' as LocalDate,
            gyldigTilOgMed: null,
            erHistorisk: false
        }
    ],
    adressebeskyttelse: [
        {
            kode: AdresseBeskyttelse.KODE6,
            beskrivelse: 'Ingen'
        }
    ],
    sikkerhetstiltak: [
        {
            type: SikkerhetstiltakType.TFUS,
            gyldigFraOgMed: '2005-02-13' as LocalDate,
            gyldigTilOgMed: '2030-02-15' as LocalDate
        }
    ],
    erEgenAnsatt: EgenAnsatt.JA,
    personstatus: [
        {
            kode: PersonStatus.BOSATT,
            beskrivelse: 'Bosatt i Norge'
        }
    ],
    sivilstand: [
        {
            type: {
                kode: SivilstandType.GIFT,
                beskrivelse: 'Gift'
            },
            gyldigFraOgMed: '2005-12-12' as LocalDate
        }
    ],
    foreldreansvar: [
        {
            ansvar: 'felles',
            ansvarlig: {
                fornavn: 'Test',
                etternavn: 'Testesen',
                mellomnavn: null
            },
            ansvarsubject: {
                fornavn: 'Barn',
                etternavn: 'Barnesen',
                mellomnavn: null
            }
        }
    ],
    deltBosted: [
        {
            startdatoForKontrakt: '2000-10-10' as LocalDate,
            sluttdatoForKontrakt: '2005-05-07' as LocalDate,
            adresse: {
                linje1: `Adresseveien 1`,
                linje2: '0000 Aremark',
                linje3: null
            }
        }
    ],
    dodsbo: [
        {
            adressat: {
                advokatSomAdressat: {
                    kontaktperson: {
                        fornavn: 'Advokat',
                        mellomnavn: null,
                        etternavn: 'Navnesen'
                    },
                    organisasjonsnavn: 'Advokatkontoret Aremark',
                    organisasjonsnummer: '01234567891011'
                },
                organisasjonSomAdressat: null,
                personSomAdressat: null
            },
            adresse: {
                linje1: 'Elgelia 20, 0000 Aremark',
                linje2: null,
                linje3: null
            },
            registrert: '2010-02-02' as LocalDate,
            skifteform: Skifteform.OFFENTLIG
        }
    ],
    fullmakt: [
        {
            motpartsPersonident: '123456789',
            motpartsPersonNavn: {
                fornavn: 'Navn',
                mellomnavn: null,
                etternavn: 'Navnesen'
            },
            motpartsRolle: FullmaktsRolle.FULLMEKTIG,
            omraade: ['*'],
            gyldigFraOgMed: '2015-01-01' as LocalDate,
            gyldigTilOgMed: '2017-12-12' as LocalDate
        }
    ],
    vergemal: [
        {
            ident: '21042900076',
            navn: {
                fornavn: 'Simen',
                mellomnavn: null,
                etternavn: 'Solli'
            },
            vergesakstype: 'voksen',
            omfang: 'personligeOgOekonomiskeInteresser',
            embete: 'Fylkesmannen i Troms og Finnmark',
            gyldighetstidspunkt: '2016-03-27' as LocalDate,
            opphoerstidspunkt: '2017-06-30' as LocalDate
        }
    ],
    tilrettelagtKommunikasjon: {
        talesprak: [
            {
                kode: 'NO',
                beskrivelse: 'Norsk'
            }
        ],
        tegnsprak: [
            {
                kode: 'NO',
                beskrivelse: 'Norsk'
            }
        ]
    },
    telefonnummer: [
        {
            retningsnummer: {
                kode: '+47',
                beskrivelse: 'retningsnummer er norsk'
            },
            identifikator: '99009900',
            sistEndret: '2017-10-10T10:15:30' as LocalDateTime,
            sistEndretAv: 'TESTFAMILIEN AREMARK',
            prioritet: 1
        }
    ],
    kontaktOgReservasjon: {
        personident: '10108000398',
        reservasjon: 'true',
        epostadresse: {
            value: 'epost@nav.no',
            sistOppdatert: null,
            sistVerifisert: '2013-01-01' as LocalDate
        },
        mobiltelefonnummer: {
            value: '90000000',
            sistOppdatert: null,
            sistVerifisert: null
        }
    },
    bankkonto: {
        kontonummer: '1234567890',
        banknavn: 'DNB ASA',
        sistEndret: '2006-03-15T00:00:00' as LocalDateTime,
        sistEndretAv: '1010800 BD03',

        bankkode: null,
        swift: 'DNBANOKKXXX',
        landkode: null,
        adresse: {
            linje1: 'Bankveien 1,',
            linje2: '0357 Bankestad',
            linje3: null
        },
        valuta: {
            kode: 'NOK',
            beskrivelse: 'Norske kroner'
        }
    },
    forelderBarnRelasjon: barnAremark
};
