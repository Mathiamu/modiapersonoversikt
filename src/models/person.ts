import { Kodeverk } from './kodeverk';

export interface Person {
    navn: Navn;
    kjønn: Kjønn;
    geografiskTilknytning?: string;
    fødselsnummer: string;
    alder: number;
    diskresjonskode?: string;
    bankkonto?: Bankkonto;
    statsborgerskap?: string;
    personstatus: Bostatus;
    sivilstand: Kodeverk;
    familierelasjoner: Familierelasjon[];
}

export interface Navn {
    sammensatt: string;
    fornavn: string;
    mellomnavn: string;
    etternavn: string;
}

export interface Bankkonto {
    erNorskKonto: boolean;
    kontonummer: number;
    bank: string;
    sistEndret: string;
    sistEndretAv: string;
}

export interface Familierelasjon {
    harSammeBosted: boolean;
    rolle: Relasjonstype;
    tilPerson: {
        navn: Navn;
        alder: number;
        fødselsnummer: string;
        personstatus: Bostatus;
        kjonn: Kjønn
    };
}

export interface Bostatus {
    dødsdato?: string;
    bostatus?: string;
}

export enum BostatusTyper {
    Død = 'DØD',
    Utvandret = 'UTVA'
}

export enum Relasjonstype {
    Barn = 'BARN',
    Samboer = 'SAMBOER',
    Partner = 'PARTNER',
    Ektefelle = 'EKTE',
    Gift = 'GIFT'
}

export enum Sivilstand {
    Gift = 'GIFT',
    Ugift = 'UGIFT'
}

export enum Kjønn {
    Mann = 'M',
    Kvinne = 'K'
}

export function erDød(personstatus: Bostatus) {
    return personstatus.dødsdato || personstatus.bostatus === BostatusTyper.Død;
}

export function getBarnUnder21(familierelasjoner: Familierelasjon[]) {
    return familierelasjoner
        .filter(relasjon => relasjon.rolle === Relasjonstype.Barn)
        .filter(relasjon => relasjon.tilPerson.alder <= 21);
}

export function getPartner(person: Person) {
    const aktuelleRelasjoner = [
        Relasjonstype.Ektefelle,
        Relasjonstype.Gift,
        Relasjonstype.Partner,
        Relasjonstype.Samboer
    ];
    return person.familierelasjoner.find(relasjon => aktuelleRelasjoner.includes(relasjon.rolle));
}
