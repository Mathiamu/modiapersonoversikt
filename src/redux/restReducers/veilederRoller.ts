import { createActionsAndReducerDeprecated } from './deprecatedRestResource';
import { fetchVeilederRoller } from '../../api/veileder-roller-api';

const { reducer, action, actionNames } = createActionsAndReducerDeprecated('veileder_roller');

export function getVeilederRoller() {
    return action(() => fetchVeilederRoller());
}

export { actionNames as veilederRollerReducerActionNames };
export default reducer;
