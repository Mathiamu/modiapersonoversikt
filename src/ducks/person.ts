import { Person } from '../models/person';
import { Action, Dispatch } from 'redux';
import { FetchError, FetchSuccess, RestActions, STATUS } from './utils';

export interface PersonState {
    data: Person;
    status: STATUS;
    error: string | null;
}

const mockPerson = {
    fornavn: 'Aremark',
    etternavn: 'Testfamilien',
    fodselsnummer: '10108000398'
};

const reducerName = 'person/';

const actionNames = {
    PENDING: reducerName + STATUS.PENDING,
    OK: reducerName + STATUS.OK,
    ERROR: reducerName + STATUS.ERROR
};

const initialState: PersonState = {
    data: mockPerson,
    status: STATUS.NOT_STARTED,
    error: null
};

function mockFetch(fodselsnummer: string) {
    return new Promise((resolve, reject) => {
        setTimeout(() => { resolve(mockPerson); }, 2500);
    });
}

export function hentPerson(fodselsnummer: string) {
    return (dispatch: Dispatch<Action>) => {
        dispatch({ type: actionNames.PENDING});
        return mockFetch(fodselsnummer)
            .then((person) => {
                dispatch({type: actionNames.OK, data: person});
                return person;
            })
            .catch((err) => {
                dispatch({type: actionNames.ERROR, error: err});
            });
    };
}

export default function reducer(state: PersonState = initialState, action: RestActions<Person>): PersonState {
    switch (action.type) {
        case actionNames.PENDING:
            return {...state, status: STATUS.PENDING };
        case actionNames.OK:
            return {...state, status: STATUS.OK, data: (<FetchSuccess<Person>> action).data};
        case actionNames.ERROR:
            return {...state, status: STATUS.ERROR, error: (<FetchError> action).error};
        default:
            return state;
    }
}