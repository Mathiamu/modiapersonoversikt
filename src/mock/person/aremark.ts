import { Diskresjonskoder } from '../../constants';
import { Person } from '../../models/person';

export const aremark: Person = {
    fødselsnummer: '10108000398',
    kjønn: 'M',
    geografiskTilknytning: '0118',
    alder: 42,
    navn: {
        sammensatt: 'AREMARK TESTFAMILIEN',
        fornavn: 'AREMARK',
        mellomnavn: '',
        etternavn: 'TESTFAMILIEN',
    },
    diskresjonskode: Diskresjonskoder.FORTROLIG_ADRESSE,
    statsborgerskap: 'NORSK',
    status: {
        dødsdato: undefined,
        bostatus: undefined
    },
    sivilstand: {
        value: 'GIFT',
        beskrivelse: 'Gift'
    },
    familierelasjoner: []
};