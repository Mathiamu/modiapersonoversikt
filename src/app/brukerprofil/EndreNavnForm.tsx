import * as React from 'react';
import { ChangeEvent, FormEvent } from 'react';
import { Action } from 'history';
import { connect, Dispatch } from 'react-redux';
import styled from 'styled-components';

import Input from 'nav-frontend-skjema/lib/input';
import Undertittel from 'nav-frontend-typografi/lib/undertittel';
import KnappBase from 'nav-frontend-knapper';
import AlertStripe from 'nav-frontend-alertstriper';

import { STATUS } from '../../redux/utils';
import { EndreNavnRequest } from '../../redux/brukerprofil/endreNavnRequest';
import { BostatusTyper, Person } from '../../models/person/person';
import { AppState } from '../../redux/reducer';
import { endreNavn, reset } from '../../redux/brukerprofil/endreNavn';
import { VeilederRoller } from '../../models/veilederRoller';
import { FormKnapperWrapper } from './BrukerprofilForm';
import RequestTilbakemelding from './kontaktinformasjon/RequestTilbakemelding';
import { erDnummer } from '../../utils/fnr-utils';

const InfomeldingWrapper = styled.div`
  margin-top: 1em;
  margin-bottom: 1em;
`;

const ENTER_KEY_PRESS = 13;

interface State {
    fornavn: {
        input: string;
        feilmelding: string | undefined;
    };
    mellomnavn: {
        input: string;
        feilmelding: string | undefined;
    };
    etternavn: {
        input: string;
        feilmelding: string | undefined;
    };
    formErEndret: boolean;
}

interface DispatchProps {
    endreNavn: (request: EndreNavnRequest) => void;
    resetEndreNavnReducer: () => void;
}

interface StateProps {
    status: STATUS;
}

interface OwnProps {
    person: Person;
    veilederRoller?: VeilederRoller;
}

type Props = DispatchProps & StateProps & OwnProps;

class EndreNavnForm extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.state = this.initialState(props);

        this.fornavnInputChange = this.fornavnInputChange.bind(this);
        this.mellomnavnInputChange = this.mellomnavnInputChange.bind(this);
        this.etternavnInputChange = this.etternavnInputChange.bind(this);
        this.tilbakestillForm = this.tilbakestillForm.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    initialState(props: Props) {
        return {
            fornavn: {
                input: props.person.navn.fornavn || '',
                feilmelding: undefined
            },
            mellomnavn: {
                input: props.person.navn.mellomnavn || '',
                feilmelding: undefined
            },
            etternavn: {
                input: props.person.navn.etternavn || '',
                feilmelding: undefined
            },
            formErEndret: false
        };
    }

    componentWillUnmount() {
        this.props.resetEndreNavnReducer();
    }

    fornavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            fornavn: {
                input: event.target.value.toUpperCase(),
                feilmelding: undefined
            },
            formErEndret: true
        });
    }

    mellomnavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            mellomnavn: {
                feilmelding: undefined,
                input: event.target.value.toUpperCase()
            },
            formErEndret: true
        });
    }

    etternavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            etternavn: {
                feilmelding: undefined,
                input: event.target.value.toUpperCase()
            },
            formErEndret: true
        });
        event.preventDefault();
    }

    brukersNavnKanEndres() {
        if (erDnummer(this.props.person.fødselsnummer)) {
            return true;
        } else if (this.props.person.personstatus === BostatusTyper.Utvandret) {
            return true;
        }

        return false;
    }

    navnErEndret() {
        const fornavnErEndret = this.state.fornavn.input !== this.props.person.navn.fornavn;
        const mellomnavnErEndret = this.state.mellomnavn.input !== this.props.person.navn.mellomnavn;
        const etternavnErEndret = this.state.etternavn.input !== this.props.person.navn.etternavn;
        return fornavnErEndret || mellomnavnErEndret || etternavnErEndret;
    }

    tilbakestillForm(event: React.MouseEvent<HTMLButtonElement>) {
        this.setState(this.initialState(this.props));
        this.props.resetEndreNavnReducer();
        event.preventDefault();
    }

    validerFornavn() {
        if (this.state.fornavn.input.trim().length === 0) {
            this.setState({
                fornavn: {
                    ...this.state.fornavn,
                    feilmelding: 'Fornavn kan ikke være tomt'
                }
            });
            return false;
        }
        return true;
    }

    validerEtternavn() {
        if (this.state.etternavn.input.trim().length === 0) {
            this.setState({
                etternavn: {
                    ...this.state.etternavn,
                    feilmelding: 'Fornavn kan ikke være tomt'
                }
            });
            return false;
        }
        return true;
    }

    handleSubmit(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();

        if (!this.validerFornavn() || !this.validerEtternavn()) {
            return;
        }

        this.setState({
            formErEndret: false
        });

        this.props.endreNavn({
            fødselsnummer: this.props.person.fødselsnummer,
            fornavn: this.state.fornavn.input,
            mellomnavn: this.state.mellomnavn.input,
            etternavn: this.state.etternavn.input
        });
    }

    harVeilderPåkrevdRolle() {
        if (!this.props.veilederRoller) {
            return false;
        }
        return this.props.veilederRoller.roller.includes('0000-GA-BD06_EndreNavn');
    }

    potensiellInfomelding() {
        if (!this.harVeilderPåkrevdRolle()) {
            return (
                <InfomeldingWrapper>
                    <AlertStripe
                        type={'info'}
                    >
                        Du har ikke nødvendig rolle for å endre navn.
                    </AlertStripe>
                </InfomeldingWrapper>
            );
        }

        if (!this.brukersNavnKanEndres()) {
            return (
                <InfomeldingWrapper>
                    <AlertStripe
                        type={'info'}
                    >
                        Bruker har ikke D-nummer eller er ikke utvandret. Du kan derfor ikke endre navnet.
                    </AlertStripe>
                </InfomeldingWrapper>
            );
        }

        return null;
    }

    render() {
        const kanEndreNavn = this.harVeilderPåkrevdRolle() && this.brukersNavnKanEndres();
        const infomelding = this.potensiellInfomelding();
        return (
            <form
                onSubmit={this.handleSubmit}
                onKeyPress={(event) => {
                    if (event.which === ENTER_KEY_PRESS) {
                        this.handleSubmit(event);
                    }
                }}
            >
                <Undertittel>Navn</Undertittel>
                {infomelding}
                <Input
                    label="Fornavn"
                    value={this.state.fornavn.input}
                    onChange={this.fornavnInputChange}
                    disabled={!kanEndreNavn}
                    feil={this.state.fornavn.feilmelding ? {feilmelding: this.state.fornavn.feilmelding} : undefined}
                />
                <Input
                    label="Mellomnavn"
                    value={this.state.mellomnavn.input}
                    onChange={this.mellomnavnInputChange}
                    disabled={!kanEndreNavn}
                    feil={this.state.mellomnavn.feilmelding ?
                        {feilmelding: this.state.mellomnavn.feilmelding} :
                        undefined}
                />
                <Input
                    label="Etternavn"
                    value={this.state.etternavn.input}
                    onChange={this.etternavnInputChange}
                    disabled={!kanEndreNavn}
                    feil={this.state.etternavn.feilmelding ?
                        {feilmelding: this.state.etternavn.feilmelding} :
                        undefined}
                />
                <FormKnapperWrapper>
                    <KnappBase
                        type="standard"
                        onClick={this.tilbakestillForm}
                        disabled={!kanEndreNavn || !this.state.formErEndret}
                    >
                        Avbryt
                    </KnappBase>
                    <KnappBase
                        type="hoved"
                        spinner={this.props.status === STATUS.PENDING}
                        disabled={!kanEndreNavn || !this.state.formErEndret}
                        autoDisableVedSpinner={true}
                    >
                        Endre navn
                    </KnappBase>
                </FormKnapperWrapper>
                {!this.state.formErEndret
                    ? (<RequestTilbakemelding
                        status={this.props.status}
                        onSuccess={'Navnet ble endret. Det kan ta noen minutter før endringene blir synlig.'}
                        onError={'Det skjedde en feil ved endring av navn.'}
                    />)
                    : null
                }
            </form>

        );
    }
}

const mapStateToProps = (state: AppState): StateProps => {
    return ({
        status: state.endreNavn.status
    });
};

function mapDispatchToProps(dispatch: Dispatch<Action>): DispatchProps {
    return {
        endreNavn: (request: EndreNavnRequest) => dispatch(endreNavn(request)),
        resetEndreNavnReducer: () => dispatch(reset())
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(EndreNavnForm);
