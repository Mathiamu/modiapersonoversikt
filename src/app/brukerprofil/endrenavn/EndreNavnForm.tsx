import * as React from 'react';
import { ChangeEvent, FormEvent } from 'react';
import { Action } from 'history';
import { connect, Dispatch } from 'react-redux';

import Input from 'nav-frontend-skjema/lib/input';
import Undertittel from 'nav-frontend-typografi/lib/undertittel';
import KnappBase from 'nav-frontend-knapper';

import { STATUS } from '../../../redux/utils';
import { EndreNavnRequest } from '../../../redux/brukerprofil/endreNavnRequest';
import { Person } from '../../../models/person/person';
import { AppState } from '../../../redux/reducer';
import { endreNavn, reset } from '../../../redux/brukerprofil/endreNavn';
import { VeilederRoller } from '../../../models/veilederRoller';
import { FormKnapperWrapper } from '../BrukerprofilForm';
import RequestTilbakemelding from '../kontaktinformasjon/RequestTilbakemelding';
import { brukersNavnKanEndres, validerNavn, veilederHarPåkrevdRolle } from './endrenavn-utils';
import Infomelding from './Infomelding';

const ENTER_KEY_PRESS = 13;

function ignoreEnter(event: React.KeyboardEvent<HTMLInputElement>) {
    if (event.which === ENTER_KEY_PRESS) {
        event.preventDefault();
    }
}

interface State {
    fornavn: {
        input: string;
        feilmelding: string | undefined;
    };
    mellomnavn: {
        input: string;
        feilmelding: string | undefined;
    };
    etternavn: {
        input: string;
        feilmelding: string | undefined;
    };
    formErEndret: boolean;
}

interface DispatchProps {
    endreNavn: (request: EndreNavnRequest) => void;
    resetEndreNavnReducer: () => void;
}

interface StateProps {
    status: STATUS;
}

interface OwnProps {
    person: Person;
    veilederRoller?: VeilederRoller;
}

type Props = DispatchProps & StateProps & OwnProps;

class EndreNavnForm extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.state = this.initialState(props);

        this.fornavnInputChange = this.fornavnInputChange.bind(this);
        this.mellomnavnInputChange = this.mellomnavnInputChange.bind(this);
        this.etternavnInputChange = this.etternavnInputChange.bind(this);
        this.tilbakestillForm = this.tilbakestillForm.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    initialState(props: Props) {
        return {
            fornavn: {
                input: props.person.navn.fornavn || '',
                feilmelding: undefined
            },
            mellomnavn: {
                input: props.person.navn.mellomnavn || '',
                feilmelding: undefined
            },
            etternavn: {
                input: props.person.navn.etternavn || '',
                feilmelding: undefined
            },
            formErEndret: false
        };
    }

    componentWillUnmount() {
        this.props.resetEndreNavnReducer();
    }

    fornavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            fornavn: {
                input: event.target.value.toUpperCase(),
                feilmelding: undefined
            },
            formErEndret: true
        });
    }

    mellomnavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            mellomnavn: {
                feilmelding: undefined,
                input: event.target.value.toUpperCase()
            },
            formErEndret: true
        });
    }

    etternavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            etternavn: {
                feilmelding: undefined,
                input: event.target.value.toUpperCase()
            },
            formErEndret: true
        });
        event.preventDefault();
    }

    navnErEndret() {
        const fornavnErEndret = this.state.fornavn.input !== this.props.person.navn.fornavn;
        const mellomnavnErEndret = this.state.mellomnavn.input !== this.props.person.navn.mellomnavn;
        const etternavnErEndret = this.state.etternavn.input !== this.props.person.navn.etternavn;
        return fornavnErEndret || mellomnavnErEndret || etternavnErEndret;
    }

    tilbakestillForm(event: React.MouseEvent<HTMLButtonElement>) {
        this.setState(this.initialState(this.props));
        this.props.resetEndreNavnReducer();
        event.preventDefault();
    }

    validerFornavn() {
        try {
            validerNavn(this.state.fornavn.input);
        } catch (error) {
            this.setState({
                fornavn: {
                    ...this.state.fornavn,
                    feilmelding: error.toString()
                }
            });
            throw error;
        }
    }

    validerEtternavn() {
        try {
            validerNavn(this.state.etternavn.input);
        } catch (error) {
            this.setState({
                etternavn: {
                    ...this.state.etternavn,
                    feilmelding: error.toString()
                }
            });
            throw error;
        }
    }

    validerInput() {
        let validInput = true;

        try {
            this.validerFornavn();
        } catch {
            validInput = false;
        }

        try {
            this.validerEtternavn();
        } catch {
            validInput = false;
        }

        return validInput;
    }

    handleSubmit(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();

        if (!this.validerInput()) {
            return;
        }

        this.setState({
            formErEndret: false
        });

        this.props.endreNavn({
            fødselsnummer: this.props.person.fødselsnummer,
            fornavn: this.state.fornavn.input,
            mellomnavn: this.state.mellomnavn.input,
            etternavn: this.state.etternavn.input
        });
    }

    render() {
        const kanEndreNavn = veilederHarPåkrevdRolle(this.props.veilederRoller) &&
            brukersNavnKanEndres(this.props.person);
        return (
            <form onSubmit={this.handleSubmit}>
                <Undertittel>Navn</Undertittel>
                <Infomelding person={this.props.person} veilderRoller={this.props.veilederRoller}/>
                <Input
                    label="Fornavn"
                    value={this.state.fornavn.input}
                    onChange={this.fornavnInputChange}
                    disabled={!kanEndreNavn}
                    onKeyPress={ignoreEnter}
                    feil={this.state.fornavn.feilmelding ? {feilmelding: this.state.fornavn.feilmelding} : undefined}
                />
                <Input
                    label="Mellomnavn"
                    value={this.state.mellomnavn.input}
                    onChange={this.mellomnavnInputChange}
                    disabled={!kanEndreNavn}
                    onKeyPress={ignoreEnter}
                    feil={this.state.mellomnavn.feilmelding ?
                        {feilmelding: this.state.mellomnavn.feilmelding} :
                        undefined}
                />
                <Input
                    label="Etternavn"
                    value={this.state.etternavn.input}
                    onChange={this.etternavnInputChange}
                    disabled={!kanEndreNavn}
                    onKeyPress={ignoreEnter}
                    feil={this.state.etternavn.feilmelding ?
                        {feilmelding: this.state.etternavn.feilmelding} :
                        undefined}
                />
                <FormKnapperWrapper>
                    <KnappBase
                        type="standard"
                        onClick={this.tilbakestillForm}
                        disabled={!kanEndreNavn || !this.state.formErEndret}
                    >
                        Avbryt
                    </KnappBase>
                    <KnappBase
                        type="hoved"
                        spinner={this.props.status === STATUS.PENDING}
                        disabled={!kanEndreNavn || !this.state.formErEndret}
                        autoDisableVedSpinner={true}
                    >
                        Endre navn
                    </KnappBase>
                </FormKnapperWrapper>
                {!this.state.formErEndret
                    ? (<RequestTilbakemelding
                        status={this.props.status}
                        onSuccess={'Navnet ble endret. Det kan ta noen minutter før endringene blir synlig.'}
                        onError={'Det skjedde en feil ved endring av navn.'}
                    />)
                    : null
                }
            </form>

        );
    }
}

const mapStateToProps = (state: AppState): StateProps => {
    return ({
        status: state.endreNavn.status
    });
};

function mapDispatchToProps(dispatch: Dispatch<Action>): DispatchProps {
    return {
        endreNavn: (request: EndreNavnRequest) => dispatch(endreNavn(request)),
        resetEndreNavnReducer: () => dispatch(reset())
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(EndreNavnForm);
