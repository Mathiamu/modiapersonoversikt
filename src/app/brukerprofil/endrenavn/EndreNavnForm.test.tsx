import * as React from 'react';
import * as renderer from 'react-test-renderer';
import TestProvider from '../../../test/Testprovider';
import { aremark } from '../../../mock/person/aremark';
import EndreNavnForm from './EndreNavnForm';
import { VeilederRoller } from '../../../models/veilederRoller';

test('Viser form for endre kontaktinformasjon', () => {

    const veilederRoller: VeilederRoller = { roller: ['0000-GA-BD06_EndreNavn']};
    const endreTilrettelagtKommunikasjonForm = renderer.create(
        <TestProvider>
            <EndreNavnForm person={aremark} veilederRoller={veilederRoller} />
        </TestProvider>
    );

    expect(endreTilrettelagtKommunikasjonForm.toJSON()).toMatchSnapshot();
});
