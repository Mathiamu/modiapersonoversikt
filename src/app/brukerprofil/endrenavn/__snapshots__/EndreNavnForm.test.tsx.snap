// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viser form for endre kontaktinformasjon 1`] = `
<form
  onSubmit={[Function]}
>
  <h2
    className="typo-undertittel"
  >
    Navn
  </h2>
  <div
    className="sc-kgoBCf eizglE"
  >
    <div
      aria-live={undefined}
      className="alertstripe alertstripe--info"
    >
      <span
        aria-label="info"
        className="alertstripe__ikon"
      >
        <svg
          focusable="false"
          height="1.5em"
          kind="info-sirkel-orange"
          onClick={undefined}
          style={undefined}
          viewBox="0 0 24 24"
          width="1.5em"
        >
          <title>
            Informasjon
          </title>
          <g
            fill="#D87F0A"
            fillRule="evenodd"
          >
            <path
              d="M12 0C5.383 0 0 5.383 0 12s5.383 12 12 12 12-5.383 12-12S18.617 0 12 0zm0 22.957C5.958 22.957 1.043 18.042 1.043 12 1.043 5.958 5.958 1.043 12 1.043c6.042 0 10.957 4.915 10.957 10.957 0 6.042-4.915 10.957-10.957 10.957z"
            />
            <path
              d="M15.13 18.783h-2.608v-8.87A.522.522 0 0 0 12 9.391H9.913a.522.522 0 0 0 0 1.044h1.565v8.348H8.87a.522.522 0 0 0 0 1.043h6.26a.522.522 0 0 0 0-1.043z"
            />
            <ellipse
              cx="11.478"
              cy="5.739"
              rx="1.043"
              ry="1.043"
            />
          </g>
        </svg>
      </span>
      <span
        className="typo-normal alertstripe__tekst"
      >
        Bruker har ikke D-nummer eller er ikke utvandret. Du kan derfor ikke endre navnet.
      </span>
    </div>
  </div>
  <div
    className="skjemaelement"
  >
    <label
      className="skjemaelement__label"
      htmlFor="Fornavn"
    >
      Fornavn
    </label>
    <input
      className="skjemaelement__input input--fullbredde"
      disabled={true}
      id="Fornavn"
      name="Fornavn"
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value="AREMARK"
    />
    <div
      aria-live="assertive"
      role="alert"
    />
  </div>
  <div
    className="skjemaelement"
  >
    <label
      className="skjemaelement__label"
      htmlFor="Mellomnavn"
    >
      Mellomnavn
    </label>
    <input
      className="skjemaelement__input input--fullbredde"
      disabled={true}
      id="Mellomnavn"
      name="Mellomnavn"
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value=""
    />
    <div
      aria-live="assertive"
      role="alert"
    />
  </div>
  <div
    className="skjemaelement"
  >
    <label
      className="skjemaelement__label"
      htmlFor="Etternavn"
    >
      Etternavn
    </label>
    <input
      className="skjemaelement__input input--fullbredde"
      disabled={true}
      id="Etternavn"
      name="Etternavn"
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value="TESTFAMILIEN"
    />
    <div
      aria-live="assertive"
      role="alert"
    />
  </div>
  <div
    className="sc-chPdSV gopUKn"
  >
    <button
      className="knapp knapp--standard knapp--disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Avbryt
    </button>
    <button
      className="knapp knapp--hoved knapp--disabled"
      disabled={true}
      type="submit"
    >
      Endre navn
    </button>
  </div>
</form>
`;
